subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchableHierarchies.java

```diff
@@ -48,18 +55,28 @@ public void registerWatchableHierarchy(File watchableHierarchy, SnapshotHierarch
         }
         if (!watchableHierarchies.contains(watchableHierarchyAbsolutePath)) {
             checkThatNothingExistsInNewWatchableHierarchy(watchableHierarchyAbsolutePath, root);
+            recentlyUsedHierarchies.addFirst(watchableHierarchy);
+            watchableHierarchies = watchableHierarchies.plus(watchableHierarchy);
+        } else {
+            recentlyUsedHierarchies.remove(watchableHierarchy);
+            recentlyUsedHierarchies.addFirst(watchableHierarchy);
         }
-        watchableHierarchies = watchableHierarchies.plus(watchableHierarchy);
+        LOGGER.info("Now considering {} as hierarchies to watch", recentlyUsedHierarchies);
     }
 
     @CheckReturnValue
-    public SnapshotHierarchy removeUnwatchedSnapshots(Stream<File> watchedHierarchies, SnapshotHierarchy root, Invalidator invalidator) {
-        this.watchableHierarchies = DefaultFileHierarchySet.of(watchedHierarchies::iterator);
+    public SnapshotHierarchy removeUnwatchedSnapshots(Predicate<File> isWatchedHierarchy, SnapshotHierarchy root, Invalidator invalidator) {
+        recentlyUsedHierarchies.removeIf(hierarchy -> !isWatchedHierarchy.test(hierarchy));
+        this.watchableHierarchies = DefaultFileHierarchySet.of(recentlyUsedHierarchies);
```

Comment:
Do we only need to rebuild `watchableHierarchies ` if `removeIf()` returned `true`, and/or if `currentRoot != root`?

Are there other ways we can short-circuit here?

Reply:
We only need to rebuild watchable hierarchies if `removeIf` returns `true` or the size of `recentlyUsedHierarchies` is above the limit. I don't see any other ways to short-circuit here.

I don't think it is worth to short-circuit rebuilding `watchableHierarchies` since we only do this once during the build and the number of `watchableHierarchies` is small.